---
title: "Asdia, predicting IBD from structure of the gut microbiome"
author: "Clovis Galiez"
output: html_document
---

```{r setup, include=FALSE}
set.seed(42)
knitr::opts_chunk$set(echo = TRUE)
```


We want to predict the IBD clinical status of patients given the abundance of microbial species living in their gut. curatedMetagenomicData package allows to retrieve a matrix of patients together with their gut's microbial composition. Each patient is assigned to a clinical status.

We focus here on the metagenomic study by Nielsen H.B. et al, published in 2014.


## Load/install packages
```{r libs, include=FALSE}
# keep this commented, it's just for reproducibility
#install.packages('curatedMetagenomicData')
#source('http://bioconductor.org/biocLite.R')
#biocLite('phyloseq')
#library(phyloseq)
#

library(ggplot2)
```

## Load the data
** Important Here ** : if you cannot execute the code just below (it requires some packages to be installed), you can load the pre-fetached data instead (next block) from the file ibdStart.RData.

```{r pressure, echo=FALSE}
library(curatedMetagenomicData)
###########################
# define study considered #
###########################
study = "NielsenHB_2014"
  # create output dir
output.dir = paste("./", study, sep = "")
dir.create(output.dir, showWarnings = F)

##########################
# extract taxonomic data #
##########################
db = "metaphlan_bugs_list.stool"
  #-----------#
  # load data #
  #-----------#
cat("\t-loading taxonomic data data\n")
data = curatedMetagenomicData( paste(study, db, sep = "."), dryrun = FALSE)
data = data[[1]]
  # get expression 
X = exprs(data)
X = t(X)
  # get covariates
meta = data@phenoData@data
status = meta$study_condition
status = factor(status)
names_of_species = colnames(X)
```


```{r}
# If you cannot run the code above, use the pre-fetched data:
load("ibdStart.RData")
```


Take a look to your data and understand what's in there.
```{r}
# The most important variables for you are:
dim(X)
head(status)
head(names_of_species)
```

## Predicting IBD


### Identify important bacterial species

Find potential correlations between abundances of bacterial species and IBD.

```{r explainatory vars}

```

What AUC ROC can you reach with the most correlated species?
Is it the presence or the abscence of this species that is a marker of IBD?


Are you satisfied? Check quickly with other people around and make a small litterature search.
Give a reason why this species is marker of the IBD, but may not be relevant for the predicting IBD.

Clean your data according to your findings in order to have relevant prediction from the abundance data.

What is the new best correlated species. What p-value it corresponds to?

Now you should realize that to get a biomarker of IBD, one needs not only the measure of 1 single species but the combination of several species.


## Learning a predictor




### Logistic regression


Use the package glmnet to use a logistic regression for classification.

Think beforehand of what potential trap you should you be aware of?
How can you avoid it?


```{r}
library(glmnet)

```

What species you find relevant for predicting IBD? Are they in accordance with the litterature?

### Bonus: Neural networks

#### Load/intsall keras package
```{r}
#install.packages("keras")
library(keras)
#install_keras()
```


#### Train a neural network using keras to predict IBD. 

Play around with the hyper parameters of your neural network that to predict the IBD status from the species abundances.

Ask yourself:
Is it overfitting?
Do you get any idea of which species is associated to IBD? 

